### PR TITLE
Add support for PATCH in viaduct

### PR DESCRIPTION
**Pull request by leplatrem**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Review by tarikeshaq — COMMENTED** [review]
Saw this passing by 😄 Thank you for doing this!!

The CI failure is probably because of the glean submodule change (which I imagine is by accident? I think we are already at a later version on main)

Beyond that, a change to [the .proto file](https://github.com/mozilla/application-services/blob/main/components/viaduct/src/fetch_msg_types.proto#L13-L22) to include the method type, running  `cargo run --bin protobuf-gen tools/protobuf_files.toml` and adding the method type conversion to [the android code](https://github.com/mozilla/application-services/blob/main/components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt#L118-L129) should be all!

As a side note, soon (probably sometime in the new year?) I'm planning to wrap up #4741. It should have zero impact on our Rust consumers (and it's actually probably not gonna be a breaking change to any of our consumes at all) but it will have change the internals of how the FFI logic works

**Comment by tarikeshaq** [comment]
eeeh looks like the android `concept-fetch` we use doesn't have the `PATCH` method either 😕 https://github.com/mozilla-mobile/android-components/blob/c32c96bdde62f62eb3088c289bf1eeaa424abc5d/components/concept/fetch/src/main/java/mozilla/components/concept/fetch/Request.kt#L119-L127 I'll leave this to @mhammond, not sure what that means (i.e is it a simple change to android components, or is there more to this)

**Comment by mhammond** [comment]
I'm fairly sure this will be a trivial change to a-c -although I'm not completely sure anything would break there? @leplatrem, are you testing this in the context of a-c/fenix?

**Comment by leplatrem** [comment]
Thank you for your inputs! :pray: 

> I'm fairly sure this will be a trivial change to a-c -although I'm not completely sure anything would break there? 

I made a try but it seems that the `PATCH` method isn't supported down the line:
```
java.net.ProtocolException: Invalid HTTP method: PATCH
	at java.base/java.net.HttpURLConnection.setRequestMethod(HttpURLConnection.java:487)
``` 
I'm a bit surprised TBH, but the docs are aligned https://download.java.net/java/early_access/loom/docs/api/java.base/java/net/HttpURLConnection.html#setRequestMethod(java.lang.String)

I could use a `GET` + `PUT` with ETag headers as a workaround. 

> @leplatrem, are you testing this in the context of a-c/fenix?

No, the use case is to have write methods in our Remote Settings client, mostly for unit tests setup for now. 
In the worse case, I could also put these write operations behind a feature and disable them on Android since it's not the target yet.

**Review by tarikeshaq — COMMENTED** [review]
Thanks for the changelog!!

Had a comment about the enum ordering (it's really unfortunate that it actually matters 😞, well I **think** it does 😅) otherwise as long as this builds and raises appropriate exceptions where it's not supported - this looks great!

**Comment by mhammond** [comment]
> I made a try but it seems that the `PATCH` method isn't supported down the line:

That's a runtime error, right? I guess I'd be fine with an approach of "well, don't do that then" for Android.

File: components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt [hunk 1]
```diff
@@ -125,6 +125,7 @@ internal fun convertMethod(m: MsgTypes.Request.Method): Request.Method {
         MsgTypes.Request.Method.PUT -> Request.Method.PUT
         MsgTypes.Request.Method.TRACE -> Request.Method.TRACE
         MsgTypes.Request.Method.CONNECT -> Request.Method.CONNECT
+        MsgTypes.Request.Method.PATCH -> Request.Method.PATCH
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt:128]: heh - same thought at the same time :)

[the remark this answers is not in the document]
Comment (leplatrem) [reply components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt]: I added an exception. Let me know if that's what you had in mind

File: components/viaduct/src/lib.rs [hunk 2]
```diff
@@ -30,6 +30,7 @@ pub(crate) mod msg_types {
 pub enum Method {
     Get,
     Head,
+    Patch,
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply components/viaduct/src/lib.rs]: Ok, gotcha, done :)

**Review by tarikeshaq — APPROVED** [review]
Awesome 🔥 🔥 

Tested this on fenix locally and everything works, even tested forcing a `PATCH` request and it errored out as expected with a really descriptive error message in rust:
```
[no-sentry] Network error: Java error: "fetch error: Unsupported HTTP method: PATCH"
```

**Comment by tarikeshaq** [comment]
@leplatrem I'm not sure if you have access permissions to land this, but I'm happy to land it for you if not!

**Review by mhammond — APPROVED** [review]

**Pull request merged**